### PR TITLE
WP-17438 Error handling clean up

### PR DESCRIPTION
**Pull request by jayma91**
Error handling clean up
https://varicent.atlassian.net/browse/WP-17438

[Import] Improve error handling for MsDynamics
https://varicent.atlassian.net/browse/WP-15948

notes:
- added API error handling

### CODE REVIEW

File: tap_dynamics/client.py [hunk 1]
```diff
@@ -207,7 +212,10 @@ def _make_request(self, method, endpoint, paging=False, headers=None, params=Non
         return results
 
     def get(self, endpoint, paging=False, headers=None, params=None):
-        return self._make_request("GET", endpoint, paging, headers=headers, params=params)
+        try:
+            return self._make_request("GET", endpoint, paging, headers=headers, params=params)
+        except DynamicsException as e:
+            raise SymonException(f'Import failed with the following MS Dynamics error: {str(e)}','dynamics.DynamicsApiError')
```
Comment (jayma91) [inline tap_dynamics/client.py:218]: example:
`tap_dynamics.symon_exception.SymonException: Import failed with the following MS Dynamics error: {"error":{"code":"0x80060888","message":"EntityMetadata With Id = LogicalName='account1' does not exist."}}`

File: tap_dynamics/client.py [hunk 2]
```diff
@@ -207,7 +224,22 @@ def _make_request(self, method, endpoint, paging=False, headers=None, params=Non
...
+            message, error_code = None, None
+            try:
+                error_json = e.response.json()
+                message = error_json["error"]["message"]
+                error_code = error_json["error"]["code"]
+            except:
+                pass
+
+            if message is not None and error_code is not None:
+                raise SymonException(f'Import failed with the following MS Dynamics error: (error code: {error_code}) {message}','dynamics.DynamicsApiError')
```
Comment (jayma91) [inline tap_dynamics/client.py:232]: example:
`tap_dynamics.symon_exception.SymonException: Import failed with the following MS Dynamics error: (error code: 0x80060888) EntityMetadata With Id = LogicalName='account1' does not exist.`

err doc:  https://learn.microsoft.com/en-us/power-apps/developer/data-platform/reference/web-service-error-codes
- not sure if this should be used, couldn't find matching matching error code used in the example.

File: tap_dynamics/client.py [hunk 3]
```diff
@@ -188,16 +200,21 @@ def _make_request(self, method, endpoint, paging=False, headers=None, params=Non
...
-        response = self.session.request(
-            method, full_url, headers=headers, params=params, data=data)
-
+        try:
+            response = self.session.request(
+                method, full_url, headers=headers, params=params, data=data)
+        except requests.exceptions.ConnectionError as e:
+            message = str(e)
+            if 'nodename nor servname provided, or not known' in message or 'Name or service not known' in message:
+                raise SymonException(f'Sorry, we couln\'t connect to Dynamics URL "{self.organization_uri}". Please check the Dynamics URL and try again.', 'dynamics.InvalidUrl')
```
Comment (mjdoor) [inline tap_dynamics/client.py]: Typo on couldn't

File: tap_dynamics/client.py [hunk 4]
```diff
@@ -44,25 +45,37 @@ def retry_after_wait_gen():
...
 
 # pylint: disable=missing-class-docstring
 
 
 class DynamicsQuotaExceededException(DynamicsException):
-    pass
+    def __init__(self, message=None, response=None):
+        super().__init__(message)
+        self.message = message
+        self.response = response
```
Comment (mjdoor) [inline tap_dynamics/client.py]: Not a big deal, but do we need message and response separately in these subclasses, or can we just pass them both through to the DynamicsException parent class?

**Review by mjdoor — APPROVED** [review]

**Pull request merged**